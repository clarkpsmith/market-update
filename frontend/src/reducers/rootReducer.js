import {
  REMOVE_TICKER,
  ADD_TICKER,
  UPDATE_CURR_USER,
  LOG_OUT,
  SIGN_UP,
  UPDATE_PROFILE,
  REFRESH_TICKER,
  SET_WATCHLIST_DATA,
  SEARCH_TICKER_DATA,
} from "../actions/types";

const DEFAULT_STATE = {
  currentUser: {},
  watchlistData: [],
  searchTickerData: null,
};

function rootReducer(state = DEFAULT_STATE, action) {
  switch (action.type) {
    case REMOVE_TICKER:
      return {
        ...state,
        currentUser: {
          ...state.currentUser,
          watchlist: state.currentUser.watchlist.filter(
            (t) => t !== action.ticker.symbol
          ),
        },
        watchlistData: state.watchlistData.filter(
          (t) => t.symbol !== action.ticker.symbol
        ),
      };
    case SEARCH_TICKER_DATA:
      return {
        ...state,
        searchTickerData: action.tickerData,
      };

    case ADD_TICKER:
      return {
        ...state,
        currentUser: {
          ...state.currentUser,
          watchlist: [...state.currentUser.watchlist, action.ticker.symbol],
        },
        watchlistData: [...state.watchlistData, action.ticker],
      };

    case UPDATE_CURR_USER:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case UPDATE_PROFILE:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case LOG_OUT:
      return {
        ...state,
        currentUser: {},
        watchlistData: [],
      };

    case SIGN_UP:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case SET_WATCHLIST_DATA:
      return {
        ...state,
        watchlistData: [...action.watchlistData],
      };

    case REFRESH_TICKER:
      return {
        ...state,
        watchlistData: [
          ...state.watchlistData.map((t) => {
            if (t.symbol === action.ticker.symbol) return action.ticker;
            else return t;
          }),
        ],
      };
    default:
      return state;
  }
}
export default rootReducer;
