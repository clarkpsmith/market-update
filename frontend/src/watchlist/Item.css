.Item {
  text-align: left;
  /* border-bottom: 1px solid grey; */
  border-bottom: 1px solid grey;
  padding: 10px 15px;

  height: 80px;
  color: white;
  background-color: black;
}

.Item-btn-watchlist {
  margin-left: 10px;

  padding: 1px 3px !important;
  margin-bottom: 5px !important;
}

.Item-btn-change {
  border-radius: 3px;
  /* padding-left: 10px !important;
  padding-right: 2px !important; */
  font-size: 18px !important;
}
.Item-title {
  margin-bottom: 0px;
}

.Item-info-icon {
  font-size: 8px;
  margin-left: 6px;
  color: rgb(200, 210, 236);
  transition: all 0.2s;
}

.Item-fullname {
  color: grey;
  font-size: 15px;
}

@media only screen and (max-width: 414px) {
  .Item {
    height: 65px;
    padding: 7px;
  }

  .Item-btn-watchlist {
    /* left: 85px;
    position: absolute;
    padding: 3px !important;
    font-size: 11px !important; */
  }
  .Item-title {
    font-size: 16px;
  }
  /* .Item-btn-change {
  border-radius: 3px;
  /* padding-left: 10px !important;
  padding-right: 2px !important; */
  /* font-size: 14px !important;
 */
  .Item-btn-change {
    /* padding-left: 10px !important;
  padding-right: 2px !important; */
    font-size: 15px !important;
  }
}
