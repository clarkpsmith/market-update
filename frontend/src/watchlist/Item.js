import React, { useState } from "react";
import "../stockinfo/Stockinfo.css";
import { Button } from "reactstrap";
import { addCommas } from "../helpers/abreviateLargeNums";
import "../stockinfo/Stockinfo.css";
import "./Item.css";

import {
  removeTickerFromList,
  addTickerToList,
} from "../actions/actionCreators";
import { SEARCH_TICKER_DATA } from "../actions/types";
import { useDispatch, useSelector } from "react-redux";
import { refreshTicker } from "../actions/actionCreators";
import Stockinfo from "../stockinfo/Stockinfo";

const Item = ({ ticker }) => {
  const currentUser = useSelector((store) => store.currentUser);
  const dispatch = useDispatch();
  let alreadyAdded = false;
  if (Object.keys(currentUser).length)
    alreadyAdded = currentUser.watchlist.filter((t) => t === ticker.symbol);
  const [expand, setExpand] = useState(false);
  const [added, setAdded] = useState(alreadyAdded.length !== 0);
  const [marketChange, setMarketChange] = useState({
    percent: `${ticker.regularMarketChangePercent.toFixed(2)} %`,
    market: ticker.regularMarketChange.toFixed(2),
  });
  const [percent, setPercent] = useState(true);

  function handleClick() {
    percent ? setPercent(false) : setPercent(true);
  }

  function addToWatchlist() {
    dispatch(addTickerToList(currentUser.username, ticker));
    dispatch({ type: SEARCH_TICKER_DATA, tickerData: null });
    setAdded(true);
  }

  async function remove() {
    dispatch(removeTickerFromList(currentUser.username, ticker));
  }

  function toggleExpand() {
    if (!expand) setExpand(true);
    else setExpand(false);
  }

  function refresh() {
    dispatch(refreshTicker(ticker.symbol));
  }

  let percentColor;
  marketChange.market >= 0
    ? (percentColor = "LimeGreen")
    : (percentColor = "red");
  const marketPrice = ticker.regularMarketPrice
    ? addCommas(ticker.regularMarketPrice.toFixed(2))
    : "N/A";

  if (!expand) {
    return (
      <div className="Item">
        <h4 className="Item-title">
          <span
            className="Item-title-clickable"
            title="Expand"
            onClick={toggleExpand}
          >
            {ticker.symbol}
            {/* <span className="Item-info-icon">
              <FontAwesomeIcon icon={faInfo} />
            </span> */}
          </span>
          {Object.keys(currentUser).length > 0 ? (
            <span>
              {!added ? (
                <Button
                  className="Item-btn-watchlist"
                  onClick={addToWatchlist}
                  size="sm"
                >
                  Add
                </Button>
              ) : (
                <Button
                  className="Item-btn-watchlist"
                  onClick={remove}
                  size="sm"
                >
                  Remove
                </Button>
              )}
            </span>
          ) : (
            ""
          )}

          <span>
            <span
              title="Current stock price, click to refresh"
              onClick={refresh}
              className="Stockinfo-price"
            >
              {marketPrice}
              {/* <span className="Item-info-icon">
                <FontAwesomeIcon icon={faInfo} />
              </span> */}
            </span>
          </span>
        </h4>
        <div className="Item-fullname">
          {ticker.longName}{" "}
          <span>
            <button
              onClick={handleClick}
              className="Stockinfo-daily-change Item-btn-change"
              style={{ backgroundColor: percentColor }}
            >
              {percent ? marketChange.percent : marketChange.market}
            </button>
          </span>
        </div>
      </div>
    );
  }
  if (expand) {
    return <Stockinfo ticker={ticker} />;
  }
};

export default Item;
