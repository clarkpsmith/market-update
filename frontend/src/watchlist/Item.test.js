import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Item from "./Item";
import { composeWithDevTools } from "redux-devtools-extension";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import rootReducer from "../reducers/rootReducer";
const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

const ticker = {
  language: "en-US",
  region: "US",
  quoteType: "EQUITY",
  quoteSourceName: "Nasdaq Real Time Price",
  triggerable: true,
  currency: "USD",
  firstTradeDateMilliseconds: 1277818200000,
  priceHint: 2,
  totalCash: 1.71410002e10,
  floatShares: 775105312,
  ebitda: 4551000064,
  shortRatio: 1.5,
  preMarketChange: -6.625,
  preMarketChangePercent: -0.985746,
  preMarketTime: 1620653399,
  targetPriceHigh: 1200.0,
  targetPriceLow: 67.0,
  targetPriceMean: 636.87,
  targetPriceMedian: 700.0,
  preMarketPrice: 665.455,
  heldPercentInsiders: 19.63,
  heldPercentInstitutions: 44.688,
  postMarketChangePercent: -0.8807036,
  postMarketTime: 1620684910,
  postMarketPrice: 623.5,
  postMarketChange: -5.539978,
  regularMarketChange: -43.330017,
  regularMarketChangePercent: -6.4443707,
  regularMarketTime: 1620676803,
  regularMarketPrice: 629.04,
  regularMarketDayHigh: 665.02,
  regularMarketDayRange: "627.6101 - 665.02",
  regularMarketDayLow: 627.6101,
  regularMarketVolume: 30633287,
  sharesShort: 51176090,
  sharesShortPrevMonth: 44732793,
  shortPercentFloat: 6.6,
  regularMarketPreviousClose: 672.37,
  bid: 628.79,
  ask: 624.35,
  bidSize: 10,
  askSize: 10,
  exchange: "NMS",
  market: "us_market",
  messageBoardId: "finmb_27444752",
  fullExchangeName: "NasdaqGS",
  shortName: "Tesla, Inc.",
  longName: "Tesla, Inc.",
  regularMarketOpen: 664.9,
  averageDailyVolume3Month: 35014572,
  averageDailyVolume10Day: 28429566,
  beta: 1.984491,
  fiftyTwoWeekLowChange: 476.37997,
  fiftyTwoWeekLowChangePercent: 3.120529,
  fiftyTwoWeekRange: "152.66 - 900.4",
  fiftyTwoWeekHighChange: -271.36005,
  fiftyTwoWeekHighChangePercent: -0.3013772,
  fiftyTwoWeekLow: 152.66,
  fiftyTwoWeekHigh: 900.4,
  earningsTimestamp: 1619467200,
  earningsTimestampStart: 1626778740,
  earningsTimestampEnd: 1627300800,
  trailingPE: 630.30054,
  pegRatio: 3.66,
  dividendsPerShare: 0.0,
  revenue: 3.594e10,
  priceToSales: 16.860685,
  marketState: "POST",
  epsTrailingTwelveMonths: 0.998,
  epsForward: 6.25,
  epsCurrentYear: 4.53,
  epsNextQuarter: 1.18,
  priceEpsCurrentYear: 138.86092,
  priceEpsNextQuarter: 533.0848,
  sharesOutstanding: 963329984,
  bookValue: 23.901,
  fiftyDayAverage: 688.3441,
  fiftyDayAverageChange: -59.30414,
  fiftyDayAverageChangePercent: -0.08615478,
  twoHundredDayAverage: 665.25146,
  twoHundredDayAverageChange: -36.211487,
  twoHundredDayAverageChangePercent: -0.05443278,
  marketCap: 605973053440,
  forwardPE: 100.64639,
  priceToBook: 26.318563,
  sourceInterval: 15,
  exchangeDataDelayedBy: 0,
  exchangeTimezoneName: "America/New_York",
  exchangeTimezoneShortName: "EDT",
  gmtOffSetMilliseconds: -14400000,
  esgPopulated: false,
  tradeable: false,
  symbol: "AAPL",
};

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Item ticker={ticker} />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", async () => {
  const { asFragment, queryByTestId } = await render(
    <Provider store={store}>
      <MemoryRouter>
        <Item ticker={ticker} />
      </MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
