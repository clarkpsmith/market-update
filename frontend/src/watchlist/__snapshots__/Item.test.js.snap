// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Item"
  >
    <h4
      class="Item-title"
    >
      <span
        title="Expand"
      >
        AAPL
      </span>
      <span>
        <span
          class="Stockinfo-price"
          title="Current stock price, click to refresh"
        >
          629.04
        </span>
      </span>
    </h4>
    <div
      class="Item-fullname"
    >
      Tesla, Inc. 
      <span>
        <button
          class="Stockinfo-daily-change Item-btn-change"
          style="background-color: red;"
        >
          -6.44 %
        </button>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
